### ISSUE TITLE
Running on different dataset

### ISSUE THREAD

**Issue by Harsha-Nori** [GODqinbin/DPBoost#2]
Hi @GODqinbin , 

Thank you for your paper, and for publishing code alongside it! While I can run your run_exp.py script, I'm having trouble adapting this code to any other dataset.

Does this codebase support the LightGBM's ```train``` or the scikit-learn style ```LGBMRegressor``` interfaces? I'd like to produce a LightGBM booster object or scikit-learn style estimator so that I can make predictions from the model and evaluate on my own metrics. 

I am able to successfully use lightgbm.cv(), but not lightgbm.train() or the scikit-learn style classes.

As a minimal example, when I try to adapt your code from run_exp.py to the lgb.train() method:

```python3 

    bagging_freq = 1
    params = {
        'boosting_type': 'gbdt',
        'objective': 'regression',
        'metric': 'rmse',
        'num_leaves': 31,
        'max_depth': 6,
        'learning_rate': 0.1,
        'num_iterations': n_trees,
        'my_n_trees': n_trees,
        'lambda_l2': 0.1,
        'bagging_freq': bagging_freq,
        'bagging_fraction':0.5,
        'max_bin': 255,
        'total_budget': total_budget,
        'boost_method': boost_method,
        'high_level_boost_round': 1,
        'inner_boost_round': inner_boost_round,
        'balance_partition': balance_partition,
        'geo_clip': 1,
        'verbose': -1,
    }
    data = lgb.Dataset(x,y)
    booster = lgb.train(params, data)
    booster.predict(x_test)

```

I get the following warnings: 
![image](https://user-images.githubusercontent.com/5302119/82293900-440dbe80-997b-11ea-82b7-c51ed4972973.png)

And like the warnings suggest, I keep getting the same model independent of what I set ```total_budget``` to. 

Any help would be greatly appreciated!

-Harsha

**Comment by QinbinLi** [comment]
Hi @Harsha-Nori ,

It should support LightGBM's `train` function. I have tried `lightgbm.train()` and `predict` function. It can work and return different prediction results given different total_budget (e.g., total_budgets_list=[1,10]). You can ignore the warnings. Can you provide an example that can reproduce your results?

**Comment by interpret-ml** [comment]
Hi @GODqinbin , 

Thank you for the prompt reply. I'm sure it's just a mistake on my part. Here's a minimal example -- thank you so much for your willingness to help out!

```python3
from sklearn.datasets import load_boston
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split
import lightgbm as lgb

X, y = load_boston(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)


for epsilon in [0.01, 0.1, 1, 10]:
    data = lgb.Dataset(X_train, y_train)
    dp_boost_params = { 
        'boosting_type': 'gbdt',
        'objective': 'regression',
        'metric': 'rmse',
        'num_leaves': 31,
        'max_depth': 6,
        'learning_rate': 0.1,
        'num_iterations': 50,
        'my_n_trees': 50,
        'lambda_l2': 0.1,
        'bagging_freq': 1,
        'bagging_fraction':0.5,
        'max_bin': 255,
        'boost_method': 'DPBoost_2level',
        'high_level_boost_round': 1,
        'inner_boost_round': 50,
        'balance_partition': 1,
        'geo_clip': 1,
        'verbose': -1,
        'total_budget': epsilon
    }
    dp_boost = lgb.train(dp_boost_params, data)

    print(f"At epsilon: {epsilon}, mean absolute error is: {mean_absolute_error(y_test, dp_boost.predict(X_test))}")
```

Which prints:

```
At epsilon: 0.01, mean absolute error is: 6.9074160420161
At epsilon: 0.1, mean absolute error is: 6.9074160420161
At epsilon: 1, mean absolute error is: 6.9074160420161
At epsilon: 10, mean absolute error is: 6.9074160420161
```

in my Jupyter notebook. Maybe you see an obvious error? 

Best, 
-Harsha

**Comment by Harsha-Nori** [comment]
Hi @GODqinbin , 

I'm sure it's just a silly mistake on my part. Thanks for your quick reply! Here's a minimal reproducible example:

```python3
from sklearn.datasets import load_boston
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split
import lightgbm as lgb

X, y = load_boston(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)
data = lgb.Dataset(X_train, y_train)

for epsilon in [0.1, 1, 10, 100]:
    dp_boost_params = { 
        'boosting_type': 'gbdt',
        'objective': 'regression',
        'metric': 'rmse',
        'num_leaves': 31,
        'max_depth': 6,
        'learning_rate': 0.1,
        'num_iterations': 50,
        'my_n_trees': 50,
        'lambda_l2': 0.1,
        'bagging_freq': 1,
        'bagging_fraction':0.5,
        'max_bin': 255,
        'boost_method': 'DPBoost_2level',
        'high_level_boost_round': 1,
        'inner_boost_round': 50,
        'balance_partition': 1,
        'geo_clip': 1,
        'verbose': -1,
        'total_budget': epsilon
    }
    dp_boost = lgb.train(dp_boost_params, data)

    print(f"At epsilon: {epsilon}, mean absolute error is: {mean_absolute_error(y_test, dp_boost.predict(X_test))}")
```

which prints:

```
At epsilon: 0.1, mean absolute error is: 6.9074160420161
At epsilon: 1, mean absolute error is: 6.9074160420161
At epsilon: 10, mean absolute error is: 6.9074160420161
At epsilon: 100, mean absolute error is: 6.9074160420161
```

in my Jupyter notebook. Maybe you see an obvious error? 

Best,
-Harsha

**Comment by Harsha-Nori** [comment]
Ok, I think I've isolated the problem to the ```boost_method = 'DPBoost_2level'``` parameter.

When I run the script by leaving that parameter out, I get the results you might expect:

```
At epsilon: 0.1, mean absolute error is: 156.8870132087974
At epsilon: 1, mean absolute error is: 21.836920055918306
At epsilon: 10, mean absolute error is: 6.914976557459597
At epsilon: 100, mean absolute error is: 6.345091243926308
```

However, the moment I include the ```boost_method``` parameter back in, the model becomes the same independent of the rest of the parameters. 

Is there another value I should be setting for this parameter? Thanks again in advance for your help!

-Harsha

**Comment by QinbinLi** [comment]
Hi @Harsha-Nori ,

The parameter should be set to 'DPBoost_2level'. I cannot see an obvious error in your example. Also, I can reproduce your results. Since I'm working on a deadline, I'll check it later. Thanks.

**Comment by baudoindesury** [comment]
Hi @QinbinLi ,

Thank you for the paper and the code! I was wondering if you found a solution for this issue?

Indeed, as @Harsha-Nori I am able to run your run_exp.py script and reproduce the results of your paper for the abalone dataset (Figure 2, (h)). I also succeeded to reproduce your results for the YearPredictionMSD dataset (Figure 2, (I)).

However, when I try to reproduce your results on the adult dataset (classification), I can't get the same results as the paper.

Below, how I adapted your `experiment` function in the `run_exp.py` script to output the test error instead of the mean and standard deviation.


    def experiment(dataset, total_budget, n_trees, boost_method = 'DPBoost_2level', inner_boost_round=50, balance_partition=1):

        x, y = load_svmlight_file(dataset)

        X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.5, shuffle = True)

        scaler = MinMaxScaler(feature_range=(-1,1))
        scaler.fit(y.reshape(-1,1))
        y_new = scaler.transform(y.reshape(-1,1))
        bagging_freq = 1
        params = {
            'boosting_type': 'gbdt',
            'objective': 'regression',
            'metric': 'rmse',
            'num_leaves': 31,
            'max_depth': 6,
            'learning_rate': 0.1,
            'num_iterations': n_trees,
            'my_n_trees': n_trees,
            'lambda_l2': 0.1,
            'bagging_freq': bagging_freq,
            'bagging_fraction':0.5,
            'max_bin': 255,
            'total_budget': total_budget,
            'boost_method': boost_method,
            'high_level_boost_round': 1,
            'inner_boost_round': inner_boost_round,
            'balance_partition': balance_partition,
            'geo_clip': 1,
            'verbose': -1,
        }

        data = lgb.Dataset(x,y_new.reshape(-1))

        model = lgb.train(params, data)
        y_pred_scaled = model.predict(X_test)

        y_pred = np.where(y_pred_scaled > 0, 1, -1)

        print(f"Test Error = {accuracy_score(y_pred,y_test)} for Epsilon = {total_budget}")


And here the results I got :

    Test Error = 0.237888198757764 for Epsilon = 1
    Test Error = 0.24807453416149072 for Epsilon = 2
    Test Error = 0.24 for Epsilon = 4
    Test Error = 0.24310559006211185 for Epsilon = 6
    Test Error = 0.24012422360248442 for Epsilon = 8
    Test Error = 0.24211180124223608 for Epsilon = 10

So, I was wondering if you have the code you used to produce the results for the adult dataset or if you have any idea why my code is producing wrong results?

Any help would be greatly appreciated!

Best,

Baudoin

**Comment by QinbinLi** [comment]
Hi @baudoindesury ,

You may try to add a line `'boost_from_average': False,` in your params. Also, I haven't figured out the possible bugs in `lgb.train()`. You can try `lgb.cv()`. Thanks.

**Comment by QinbinLi** [comment]
@Harsha-Nori ,

Sorry that I just found this issue. I found that the reason that the outputs are same regardless of the budget is that the size of the dataset is too small (i.e., 506). When you set the number of tree to 50, since each tree uses different subsets of the input dataset in DPBoost, the data for each tree is too small to conduct training. If you set the number of trees smaller (e.g., 10), you should be able to see different output results with different budgets.
